# Lab: Sentiment Analysis  
#  *******Data-Centric vs Model-Centric approaches




This lab gives an introduction to sentiment analysis approaches.

In this lab, we'll build a classifier for product reviews (restricted to the magazine category), like:

> Excellent! I look forward to every issue. I had no idea just how much I didn't know.  The letters from the subscribers are educational, too.

Label: ⭐️⭐️⭐️⭐️⭐️ (good)

> My son waited and waited, it took the 6 weeks to get delivered that they said it would but when it got here he was so dissapointed, it only took him a few minutes to read it.

Label: ⭐️ (bad)

We'll work with a dataset that has some issues, and we'll see how we can squeeze only so much performance out of the model by being clever about model choice, searching for better hyperparameters, etc. Then, we'll take a look at the data (as any good data scientist should), develop an understanding of the issues, and use simple approaches to improve the data. Finally, we'll see how improving the data can improve results.

## Installing software

For this lab, you'll need to install [scikit-learn](https://scikit-learn.org/) and [pandas](https://pandas.pydata.org/). If you don't have them installed already, you can install them by running the following cell:

In [9]:
!pip install scikit-learn pandas

# Loading the data

First, let's load the train/test sets and take a look at the data.

In [10]:
import pandas as pd

In [11]:
train = pd.read_csv('reviews_train.csv')
test = pd.read_csv('reviews_test.csv')

test.sample(5)

,review,label
913,Too much advertisement ! Not so much things to...,bad
293,I have loved Vogue for over 30 years!,good
429,My favorite magazine... always has good ideas ...,good
430,Great price. Love redbook,good
235,This is such a great magazine. You will find i...,good


# Training a baseline model

There are many approaches for training a sequence classification model for text data. In this lab, we're giving you code that mirrors what you find if you look up [how to train a text classifier](https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html), where we'll train an SVM on [tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) features (numeric representations of each text field based on word occurrences).

In [12]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC

In [15]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn import metrics

def evaluate(clf):
    pred = clf.predict(test['review'])
    acc = metrics.accuracy_score(test['label'], pred)
    print(f'Accuracy: {100*acc:.1f}%')

svc_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SVC(kernel='linear', C=1.0)),
])

svc_model.fit(train['review'], train['label'])

evaluate(svc_model)

Accuracy: 78.8%


In [16]:
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline

sgd_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SGDClassifier())
])

_ = sgd_clf.fit(train['review'], train['label'])

## Evaluating model accuracy

In [17]:
from sklearn import metrics

In [18]:
def evaluate(clf):
    pred = clf.predict(test['review'])
    acc = metrics.accuracy_score(test['label'], pred)
    print(f'Accuracy: {100*acc:.1f}%')

In [19]:
evaluate(sgd_clf)

Accuracy: 76.7%


## Trying another model

76% accuracy is not great for this binary classification problem. Can you do better with a different model, or by tuning hyperparameters for the SVM trained with SGD?

# Exercise 1

Can you train a more accurate model on the dataset (without changing the dataset)? You might find this [scikit-learn classifier comparison](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html) handy, as well as the [documentation for supervised learning in scikit-learn](https://scikit-learn.org/stable/supervised_learning.html).

One idea for a model you could try is a [naive Bayes classifier](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html).

You could also try experimenting with different values of the model hyperparameters, perhaps tuning them via a [grid search](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).

Or you can even try training multiple different models and [ensembling their predictions](https://scikit-learn.org/stable/modules/ensemble.html#voting-classifier), a strategy often used to win prediction competitions like Kaggle.

**Advanced:** If you want to be more ambitious, you could try an even fancier model, like training a Transformer neural network. If you go with that, you'll want to fine-tune a pre-trained model. This [guide from HuggingFace](https://huggingface.co/docs/transformers/training) may be helpful.

In [20]:
#1.1
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.svm import SVC

svc_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SVC(kernel='linear')),
])

svc_model.fit(train['review'], train['label'])

evaluate(svc_model)

Accuracy: 78.8%


I used **(SVM)**
The model achieved an improved accuracy of **78.8%**

but 78.8% accuracy still is not great for this binary classification problem.

In [21]:
#1.2

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.ensemble import RandomForestClassifier

rf_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(n_estimators=100)),
])

rf_model.fit(train['review'], train['label'])

evaluate(rf_model)

Accuracy: 84.1%


In [22]:
#1.3

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.neural_network import MLPClassifier

mlp_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MLPClassifier(max_iter=300)),
])

mlp_model.fit(train['review'], train['label'])

evaluate(mlp_model)

Accuracy: 73.8%


In [23]:
#1.4
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.ensemble import AdaBoostClassifier

ada_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', AdaBoostClassifier()),
])

ada_model.fit(train['review'], train['label'])

evaluate(ada_model)

Accuracy: 71.9%


**Traying the models again after changing the parameters**

In [24]:
#1.1

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.ensemble import RandomForestClassifier

rf_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier()),
])

rf_model.fit(train['review'], train['label'])

evaluate(rf_model)

Accuracy: 83.4%


In [25]:
#1.2

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.svm import SVC

svc_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SVC()),
])

svc_model.fit(train['review'], train['label'])

evaluate(svc_model)

Accuracy: 82.9%


In [26]:
#1.3

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.neural_network import MLPClassifier

mlp_model = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MLPClassifier()),
])

mlp_model.fit(train['review'], train['label'])

evaluate(mlp_model)

Accuracy: 76.5%


## Taking a closer look at the training data

Let's actually take a look at some of the training data:

In [27]:
train.head()

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad


Zooming in on one particular data point:

In [28]:
print(train.iloc[0].to_dict())

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}


This data point is labeled "good", but it's clearly a negative review. Also, it looks like there's some funny HTML stuff at the end.

# Exercise 2

Take a look at some more examples in the dataset. Do you notice any patterns with bad data points?

In [29]:
# show random samples
train.sample(10)

# print a few specific rows
for i in range(5):
    print(train.iloc[i].to_dict())
    print("-----")

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}
-----
{'review': "I still have not received this.  Obviously I can't review something I haven't seen.  Where is my order???????", 'label': 'bad'}
-----
{'review': '</tr>The magazine is not worth the cost of subscription.</tr>', 'label': 'good'}
-----
{'review': 'This magazine is basically ads. Kindve worthless to me really.', 'label': 'bad'}
-----
{'review': "The only thing I've recieved, so far, is the bill.\nI know this is [ or was when I last subscribed ] a ten month magazine - but so far not even the usual promotional papers.......", 'label': 'bad'}
-----


## Issues in the data

It looks like there's some funny HTML tags in our dataset, and those datapoints have nonsense labels. Maybe this dataset was collected by scraping the internet, and the HTML wasn't quite parsed correctly in all cases.

# Exercise 3

To address this, a simple approach we might try is to throw out the bad data points, and train our model on only the "clean" data.

Come up with a simple heuristic to identify data points containing HTML, and filter out the bad data points to create a cleaned training set.

In [30]:
def is_bad_data(review: str) -> bool:
    review = review.lower()
    return '<' in review or '>' in review

## Creating the cleaned training set

In [31]:
train_clean = train[~train['review'].map(is_bad_data)]

## Evaluating a model trained on the clean training set

In [32]:
from sklearn import clone

In [33]:
sgd_clf_clean = clone(sgd_clf)

In [34]:
_ = sgd_clf_clean.fit(train_clean['review'], train_clean['label'])

This model should do significantly better:

In [35]:
evaluate(sgd_clf_clean)

Accuracy: 97.0%






---




**Part 1: Function Design (Core Task)**

Task 1: Build a Model Comparison Function

You are required to implement a Python function that:

Function Requirements
Your function should:

____________________________________________
Accept:

A list (or dictionary) of machine learning models

Training data: X_train, y_train

Testing data: X_test, y_test

___________________________________________

Perform the following steps for each model:

Train the model

Make predictions on the test set

Evaluate performance

_________________________________________
Return a structured output (recommended: DataFrame or dictionary) containing:

Model name

Accuracy

Precision

Recall

F1-score

In [36]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import pandas as pd

def compare_models(models, X_train, y_train, X_test, y_test):
    results = []

    for name, model in models.items():
        model.fit(X_train, y_train)
        predictions = model.predict(X_test)

        acc = accuracy_score(y_test, predictions)
        prec = precision_score(y_test, predictions, average='weighted')
        rec = recall_score(y_test, predictions, average='weighted')
        f1 = f1_score(y_test, predictions, average='weighted')

        results.append([name, acc, prec, rec, f1])

    df = pd.DataFrame(results, columns=['Model', 'Accuracy', 'Precision', 'Recall', 'F1'])
    return df

In [37]:
models = {
    "SVM": svc_model,
    "SGD": sgd_clf,
    "Random Forest": rf_model,
    "MLP": mlp_model,
    "AdaBoost": ada_model
}

results = compare_models(models, train['review'], train['label'], test['review'], test['label'])

results

,Model,Accuracy,Precision,Recall,F1
0,SVM,0.829,0.829033,0.829,0.828996
1,SGD,0.767,0.767241,0.767,0.766948
2,Random Forest,0.836,0.837561,0.836,0.835810
3,MLP,0.747,0.747167,0.747,0.746957
4,AdaBoost,0.719,0.766451,0.719,0.705907


Random Forest achieved the best performance among the tested models, while AdaBoost had the lowest performance.